# creating a complete master dataset

In [79]:
# load all datas
import pandas as pd
import numpy as np

stores_df = pd.read_csv(
    "data/switzerland_supermarkets_clean.csv",
    dtype={"postal_code": "string"}
)

population_df = pd.read_csv(
    "data/switzerland_population_clean.csv",
    dtype={"postal_code": "string"}
)

taxpayers_df = pd.read_csv(
    "data/num_taxpayers_by_gemeinde.csv"
)

income_df = pd.read_csv(
    "data/total_income_by_gemeinde.csv"
)

postcode_municipality_df = pd.read_csv(
    "data/postal_code_municipality_mapping.csv",
    dtype={"postal_code": "string"}
)

postcode_coordinates_df = pd.read_csv(
    "data/postcode_coordinates.csv",
    dtype={"postal_code": "string"}
)


In [80]:
# Display all columns without truncation
pd.set_option("display.max_columns", None)
pd.set_option("display.width", None)

dataframes = {
    "stores_df": stores_df,
    "population_df": population_df,
    "taxpayers_df": taxpayers_df,
    "income_df": income_df,
    "postcode_municipality_df": postcode_municipality_df,
    "postcode_coordinates_df": postcode_coordinates_df
}

for name, dataframe in dataframes.items():
    print(f"\n{name}")
    print("-" * len(name))
    print(dataframe.columns.tolist())


stores_df
---------
['osm_type', 'osm_id', 'name', 'brand', 'operator', 'shop_type', 'street', 'house_number', 'postal_code', 'city', 'latitude', 'longitude', 'opening_hours', 'website', 'store_group']

population_df
-------------
['postal_code', 'population']

taxpayers_df
------------
['Kanton ID', 'Kanton / Canton', 'Gemeinde ID', 'Gemeinde / Commune', '0', "1 - 30'000", "30'001 - 40'000", "40'001 - 50'000", "50'001 - 75'000", "75'001 - 100'000", "100'001 - 200'000", "200'001 - 500'000", "500'001 - 1'000'000", "1'000'001 - u.m./et plus", 'num_taxpayers']

income_df
---------
['Kanton ID', 'Kanton / Canton', 'Gemeinde ID', 'Gemeinde / Commune', '0', "1 - 30'000", "30'001 - 40'000", "40'001 - 50'000", "50'001 - 75'000", "75'001 - 100'000", "100'001 - 200'000", "200'001 - 500'000", "500'001 - 1'000'000", "1'000'001 - u.m./et plus", 'Total_income']

postcode_municipality_df
------------------------
['postal_code', 'municipality_id', 'weight_pct', 'municipality_name', 'canton_code']

po

In [81]:
for name, dataframe in dataframes.items():
    print(f"\n{name}: shape = {dataframe.shape}")
    display(dataframe.head())


stores_df: shape = (2795, 15)


,osm_type,osm_id,name,brand,operator,shop_type,street,house_number,postal_code,city,latitude,longitude,opening_hours,website,store_group
0,node,36726161,Migros,Migros,Genossenschaft Migros Zürich,supermarket,Wiesentalstrasse,4,8730,Uznach,47.228183,8.965330,Mo-Th 07:30-19:00; Fr 07:30-20:00; Sa 07:30-17...,https://filialen.migros.ch/de/migros-supermark...,Migros
1,node,39768209,Coop,Coop,Coop Genossenschaft,supermarket,NaN,NaN,8730,Uznach,47.225154,8.969868,NaN,NaN,Coop
2,node,39947904,Coop,Coop,Coop Genossenschaft,supermarket,Bahnhofbrücke,1,8001,Zürich,47.376732,8.542161,Mo-Sa 06:00-22:00,NaN,Coop
3,node,48932835,Migros,Migros,Genossenschaft Migros Zürich,supermarket,Wengistrasse,7,8004,Zürich,47.375020,8.522895,Mo-Sa 08:00-21:00; PH off,https://filialen.migros.ch/de/migros-supermark...,Migros
4,node,70656488,Migros,Migros,Genossenschaft Migros Ostschweiz,supermarket,Zürcherstrasse,102,8406,Winterthur,47.491874,8.706448,Mo-Fr 07:30-20:00; PH off; Sa 08:00-19:00,https://filialen.migros.ch/de/migros-supermark...,Migros



population_df: shape = (3176, 2)


,postal_code,population
0,1000,4248
1,1003,6879
2,1004,31463
3,1005,12454
4,1006,15621



taxpayers_df: shape = (2148, 15)


,Kanton ID,Kanton / Canton,Gemeinde ID,Gemeinde / Commune,0,1 - 30'000,30'001 - 40'000,40'001 - 50'000,50'001 - 75'000,75'001 - 100'000,100'001 - 200'000,200'001 - 500'000,500'001 - 1'000'000,1'000'001 - u.m./et plus,num_taxpayers
0,1.0,Zürich,1,Aeugst am Albis,0.0,222.0,73.0,78.0,179.0,139.0,284.0,96.0,12.0,0.0,1087
1,1.0,Zürich,2,Affoltern am Albis,0.0,1540.0,540.0,725.0,1692.0,982.0,1126.0,210.0,11.0,0.0,6829
2,1.0,Zürich,3,Bonstetten,0.0,541.0,161.0,206.0,568.0,488.0,754.0,188.0,7.0,0.0,2914
3,1.0,Zürich,4,Hausen am Albis,4.0,429.0,132.0,213.0,440.0,285.0,457.0,117.0,6.0,0.0,2086
4,1.0,Zürich,5,Hedingen,0.0,402.0,126.0,160.0,425.0,322.0,473.0,135.0,17.0,0.0,2064



income_df: shape = (2148, 15)


,Kanton ID,Kanton / Canton,Gemeinde ID,Gemeinde / Commune,0,1 - 30'000,30'001 - 40'000,40'001 - 50'000,50'001 - 75'000,75'001 - 100'000,100'001 - 200'000,200'001 - 500'000,500'001 - 1'000'000,1'000'001 - u.m./et plus,Total_income
0,1.0,Zürich,1,Aeugst am Albis,217324.0,3225594.0,3330022.0,4136030.0,13092018.0,13644819.0,36788871.0,24176521.0,7950800.0,0.0,110358999
1,1.0,Zürich,2,Affoltern am Albis,1416064.0,28672568.0,26689741.0,42744427.0,121067589.0,83979374.0,116006928.0,48036099.0,5955601.0,0.0,475619991
2,1.0,Zürich,3,Bonstetten,0.0,8692599.0,8081305.0,11095957.0,45731404.0,47087761.0,91731834.0,39683086.0,3752700.0,0.0,258113870
3,1.0,Zürich,4,Hausen am Albis,472627.0,7333511.0,6774464.0,11250037.0,32949344.0,25297500.0,56182715.0,26350887.0,3928325.0,0.0,178368844
4,1.0,Zürich,5,Hedingen,432570.0,6232366.0,5546994.0,10573020.0,31160558.0,30005530.0,58666229.0,33505519.0,8478800.0,0.0,185857586



postcode_municipality_df: shape = (4985, 5)


,postal_code,municipality_id,weight_pct,municipality_name,canton_code
0,1000,5586,100.0,Lausanne,VD
1,1003,5586,100.0,Lausanne,VD
2,1004,5586,100.0,Lausanne,VD
3,1005,5586,100.0,Lausanne,VD
4,1006,5586,100.0,Lausanne,VD



postcode_coordinates_df: shape = (3190, 4)


,postal_code,postcode_latitude,postcode_longitude,locality_name
0,1000,46.552076,6.686622,Lausanne 25
1,1003,46.520577,6.631877,Lausanne
2,1004,46.528291,6.618649,Lausanne
3,1005,46.520897,6.642754,Lausanne
4,1006,46.512854,6.633803,Lausanne


# cleaning Postal codes

In [82]:
#formatting postalcode in all dataframe - so all postcodes have the same 4 digit text format
for dataframe in [stores_df, population_df, postcode_municipality_df, postcode_coordinates_df]:
    dataframe["postal_code"] = (
        dataframe["postal_code"]
        .astype("string")   # converts all into string
        .str.extract(r"(\d+)")[0] # extracts continuous digits and removes '.' or other alphabets
        .str.zfill(4) # adds 0 to create a clean 4 digits
    )

In [83]:
# Find valid postcodes that exist in the population dataset
valid_population_postcodes = set(population_df["postal_code"])

valid_postcode_coordinates = (postcode_coordinates_df[postcode_coordinates_df["postal_code"].isin(valid_population_postcodes)]
    .dropna(subset=["postcode_latitude","postcode_longitude"])
    .copy()
)

In [84]:
# Find stores with a missing or invalid postcode
needs_postcode_assignment = (stores_df["postal_code"].isna() | ~stores_df["postal_code"].isin(valid_population_postcodes))

print("Stores requiring postcode assignment:", needs_postcode_assignment.sum())

Stores requiring postcode assignment: 1172


In [85]:
stores_df["original_postal_code"] = (stores_df["postal_code"])

stores_df["postcode_source"] = np.where(
    needs_postcode_assignment,
    "Nearest valid population postcode",
    "OSM address"
)

In [86]:
stores_df["postal_code"].isna().sum() #stores missing postal codes

np.int64(1161)

In [87]:
# Assign missing store postcodes
# Use the nearest official postcode coordinate
from sklearn.neighbors import BallTree

#known_postcodes = postcode_coordinates_df.dropna(subset=["postcode_latitude", "postcode_longitude"]).copy()

# BallTree with haversine distance requires radians
postcode_coordinates_rad = np.radians(valid_postcode_coordinates[["postcode_latitude", "postcode_longitude"]].to_numpy())

tree = BallTree(postcode_coordinates_rad, metric="haversine")

# missing_postcode_mask = stores_df["postal_code"].isna()
#changing this because it missed few stores


store_coordinates_rad = np.radians(stores_df.loc[needs_postcode_assignment, ["latitude", "longitude"]].to_numpy())

distance, index = tree.query(store_coordinates_rad, k=1) # find closest distance
#print(distance)
stores_df.loc[needs_postcode_assignment, "postal_code"] = valid_postcode_coordinates.iloc[index.flatten()]["postal_code"].to_numpy()

# Convert angular distance into kilometres
stores_df.loc[needs_postcode_assignment, "postcode_distance_km"] = distance.flatten() * 6371 #radians to KM (angular dist x earth radius in km)


In [88]:

print("Stores still outside population postcodes:",( ~stores_df["postal_code"].isin(population_df["postal_code"])).sum())

print("\nStore totals after postcode assignment:")
print(stores_df["store_group"].value_counts().sort_index())


Stores still outside population postcodes: 0

Store totals after postcode assignment:
store_group
Aldi      240
Coop      923
Denner    724
Lidl      191
Migros    717
Name: count, dtype: int64


In [89]:
# Mark whether a postcode came from OSM or was estimated.
# Adding this before replacing the missing values:
stores_df["postcode_source"] = np.where(stores_df["postal_code"].notna(),"OSM address", "Nearest postcode coordinate")

#checking
#print(stores_df["postcode_source"].value_counts())

#print(stores_df["postcode_distance_km"].describe())

#print("Remaining missing postcodes:",stores_df["postal_code"].isna().sum())

In [90]:
# Review stores assigned unusually far away
stores_df[stores_df["postcode_distance_km"] > 4][  #Nearest coordinates are an approximation
    [
        "name",
        "store_group",
        "latitude",
        "longitude",
        "postal_code",
        "postcode_distance_km"
    ]
].sort_values("postcode_distance_km", ascending=False).head(20)

,name,store_group,latitude,longitude,postal_code,postcode_distance_km
273,Coop,Coop,46.473947,7.13323,1660,4.145117


In [91]:
# Combine income and taxpayers

In [92]:
municipality_income_df = taxpayers_df[
    [
        "Gemeinde ID",
        "Kanton / Canton",
        "Gemeinde / Commune",
        "num_taxpayers"
    ]].merge(income_df[
        [
            "Gemeinde ID",
            "Total_income"
        ]
    ],
    on="Gemeinde ID",
    how="inner",
    validate="one_to_one"
)

In [93]:
# Calculate average income per taxpayer
municipality_income_df["avg_income_per_taxpayer"] = (municipality_income_df["Total_income"] / municipality_income_df["num_taxpayers"])

municipality_income_df.replace([np.inf, -np.inf], np.nan,inplace=True)

,Gemeinde ID,Kanton / Canton,Gemeinde / Commune,num_taxpayers,Total_income,avg_income_per_taxpayer
0,1,Zürich,Aeugst am Albis,1087,110358999,101526.218031
1,2,Zürich,Affoltern am Albis,6829,475619991,69647.091961
2,3,Zürich,Bonstetten,2914,258113870,88577.168840
3,4,Zürich,Hausen am Albis,2086,178368844,85507.595398
4,5,Zürich,Hedingen,2064,185857586,90047.280039
...,...,...,...,...,...,...
2143,6806,Jura,Vendlincourt,341,17849757,52345.328446
2144,6807,Jura,Basse-Allaine,734,38730578,52766.455041
2145,6808,Jura,Clos du Doubs,786,40200279,51145.393130
2146,6809,Jura,Haute-Ajoie,680,35733362,52549.061765


In [94]:
municipality_income_df[["num_taxpayers", "Total_income", "avg_income_per_taxpayer"]].describe()

,num_taxpayers,Total_income,avg_income_per_taxpayer
count,2148.000000,2.148000e+03,2148.000000
mean,2233.921322,1.547376e+08,68888.870110
std,7236.639322,5.147586e+08,14966.469110
min,18.000000,6.914580e+05,33603.121951
25%,419.750000,2.773144e+07,59838.342207
50%,908.000000,6.297404e+07,66443.684501
75%,2149.500000,1.489599e+08,73976.897301
max,237310.000000,1.830524e+10,234227.167513


In [95]:
# Estimate postcode-level income
# Merge the municipal income with the postcode
postcode_income_df = postcode_municipality_df.merge(
    municipality_income_df[
        [
            "Gemeinde ID",
            "avg_income_per_taxpayer"
        ]
    ],
    left_on="municipality_id", # use this column on postcode_income_df
    right_on="Gemeinde ID", # use this column on postcode_municipality_df
    how="left",
    validate="many_to_one" # many postcode-mapping rows → one municipality-income row
)
postcode_income_df.head(2)

,postal_code,municipality_id,weight_pct,municipality_name,canton_code,Gemeinde ID,avg_income_per_taxpayer
0,1000,5586,100.0,Lausanne,VD,5586.0,57188.176061
1,1003,5586,100.0,Lausanne,VD,5586.0,57188.176061


In [96]:
# Calculate the weighted contribution
postcode_income_df["weighted_income"] = postcode_income_df["avg_income_per_taxpayer"]* postcode_income_df["weight_pct"] / 100

In [97]:
municipality_count = (postcode_income_df
    .groupby("postal_code")["municipality_name"]
    .nunique()
    .reset_index(name="number_of_municipalities")
)

multiple_municipality_postcodes = municipality_count[
    municipality_count["number_of_municipalities"] > 1
]

multiple_municipality_postcodes


,postal_code,number_of_municipalities
6,1008,2
11,1015,3
15,1023,2
18,1026,2
23,1031,2
...,...,...
3178,9630,3
3180,9633,4
3181,9642,2
3182,9643,2


In [98]:
# the same postcode can appear multiple times when it covers multiple municipalities
# Aggregate to one row per postcode
# estimating purchasing power. 
# because the crosswalk does not describe how a municipality’s total income is distributed among all its postcodes.
postcode_income_summary = (postcode_income_df.groupby("postal_code", as_index=False)
    .agg(
        avg_income_per_taxpayer=(
            "weighted_income",
            "sum"
        ),
        municipality_name=(
            "municipality_name",
            lambda x: ", ".join(sorted(set(x.dropna())))
        ),
        canton_code=(
            "canton_code",
            "first"
        )
    )
)
postcode_income_summary.head()

,postal_code,avg_income_per_taxpayer,municipality_name,canton_code
0,1000,57188.176061,Lausanne,VD
1,1003,57188.176061,Lausanne,VD
2,1004,57188.176061,Lausanne,VD
3,1005,57188.176061,Lausanne,VD
4,1006,57188.176061,Lausanne,VD


# merge stores and income based on postal codes

In [99]:
# finding missing stores
unmatched_stores = stores_df[~stores_df["postal_code"].isin(population_df["postal_code"])].copy()

print("Number of unmatched stores:", len(unmatched_stores))

print(unmatched_stores["store_group"].value_counts())

display(
    unmatched_stores[
        [
            "name",
            "store_group",
            "postal_code",
            "city",
            "latitude",
            "longitude",
            "postcode_source",
            "postcode_distance_km"
        ]
    ]
)

Number of unmatched stores: 0
Series([], Name: count, dtype: int64)


,name,store_group,postal_code,city,latitude,longitude,postcode_source,postcode_distance_km


In [100]:
# Count stores by postcode and brand
store_counts_df = (stores_df
    .groupby(
        ["postal_code", "store_group"]
    )
    .size()
    .unstack(fill_value=0)
    .reset_index()
)
store_counts_df

store_group,postal_code,Aldi,Coop,Denner,Lidl,Migros
0,1003,2,1,1,2,2
1,1004,1,5,1,0,3
2,1005,0,3,0,0,0
3,1006,0,2,2,0,1
4,1007,0,2,1,0,4
...,...,...,...,...,...,...
966,9630,1,1,1,1,1
967,9642,0,1,0,0,1
968,9650,0,1,0,0,0
969,9652,0,0,0,0,1


In [101]:
# Rename the columns
store_counts_df = store_counts_df.rename(
    columns={
        "Aldi": "aldi_count",
        "Coop": "coop_count",
        "Denner": "denner_count",
        "Lidl": "lidl_count",
        "Migros": "migros_count"
    }
)

# Ensure every expected column exists
store_count_columns = [
    "migros_count",
    "coop_count",
    "denner_count",
    "aldi_count",
    "lidl_count"
]

for column in store_count_columns:
    if column not in store_counts_df.columns:
        store_counts_df[column] = 0

In [102]:
# Create the business groupings
store_counts_df["external_competitor_count"] = (
    store_counts_df["coop_count"]
    + store_counts_df["aldi_count"]
    + store_counts_df["lidl_count"]
)

store_counts_df["migros_group_count"] = (
    store_counts_df["migros_count"]
    + store_counts_df["denner_count"] # Denner is a wholly owned discount subsidiary of the Migros-Genossenschafts-Bund (MGB)
)

store_counts_df

store_group,postal_code,aldi_count,coop_count,denner_count,lidl_count,migros_count,external_competitor_count,migros_group_count
0,1003,2,1,1,2,2,5,3
1,1004,1,5,1,0,3,6,4
2,1005,0,3,0,0,0,3,0
3,1006,0,2,2,0,1,2,3
4,1007,0,2,1,0,4,2,5
...,...,...,...,...,...,...,...,...
966,9630,1,1,1,1,1,3,2
967,9642,0,1,0,0,1,1,1
968,9650,0,1,0,0,0,1,0
969,9652,0,0,0,0,1,0,1


# Build a master dataset

In [103]:
population_df

,postal_code,population
0,1000,4248
1,1003,6879
2,1004,31463
3,1005,12454
4,1006,15621
...,...,...
3171,9652,781
3172,9655,363
3173,9656,665
3174,9657,715


In [104]:
# Using the population as the base because every populated postcode should remain in the opportunity analysis
# including postcodes with zero supermarkets.
analysis_df = (
    population_df
    .merge(
        postcode_income_summary,
        on="postal_code",
        how="left",
        validate="one_to_one"
    )
    .merge(
        store_counts_df,
        on="postal_code",
        how="left",
        validate="one_to_one"
    )
    .merge(
        postcode_coordinates_df,
        on="postal_code",
        how="left",
        validate="one_to_one"
    )
)

In [105]:
analysis_df.isna().sum()

postal_code                     0
population                      0
avg_income_per_taxpayer         6
municipality_name               6
canton_code                     6
aldi_count                   2205
coop_count                   2205
denner_count                 2205
lidl_count                   2205
migros_count                 2205
external_competitor_count    2205
migros_group_count           2205
postcode_latitude               6
postcode_longitude              6
locality_name                   6
dtype: int64

In [106]:
# to be safe - Fill missing store counts with zero
analysis_df[store_count_columns] = (
    analysis_df[store_count_columns]
    .fillna(0)
    .astype(int)
)

analysis_df[
    [
        "external_competitor_count",
        "migros_group_count"
    ]
] = (
    analysis_df[
        [
            "external_competitor_count",
            "migros_group_count"
        ]
    ]
    .fillna(0)
    .astype(int)
)

In [107]:
analysis_df.isna().sum() 


postal_code                  0
population                   0
avg_income_per_taxpayer      6
municipality_name            6
canton_code                  6
aldi_count                   0
coop_count                   0
denner_count                 0
lidl_count                   0
migros_count                 0
external_competitor_count    0
migros_group_count           0
postcode_latitude            6
postcode_longitude           6
locality_name                6
dtype: int64

In [108]:
#six postcodes exist in the population dataset but do not have a matching row in the postcode–municipality or postcode-coordinate datasets.
missing_location_rows = analysis_df[
    analysis_df["municipality_name"].isna()
][
    [
        "postal_code",
        "population",
        "avg_income_per_taxpayer",
        "municipality_name",
        "canton_code",
        "postcode_latitude",
        "postcode_longitude",
        "locality_name"
    ]
]

missing_location_rows

,postal_code,population,avg_income_per_taxpayer,municipality_name,canton_code,postcode_latitude,postcode_longitude,locality_name
127,1200,4319,NaN,NaN,NaN,NaN,NaN,NaN
705,2500,657,NaN,NaN,NaN,NaN,NaN,NaN
851,3000,740,NaN,NaN,NaN,NaN,NaN,NaN
1305,4000,64,NaN,NaN,NaN,NaN,NaN,NaN
1771,6000,356,NaN,NaN,NaN,NaN,NaN,NaN
2515,8000,1220,NaN,NaN,NaN,NaN,NaN,NaN


In [109]:
population_postcodes = set(
    population_df["postal_code"]
)

mapping_postcodes = set(
    postcode_municipality_df["postal_code"]
)

missing_postcodes = (
    population_postcodes - mapping_postcodes
)

print(missing_postcodes)

{'4000', '3000', '8000', '2500', '1200', '6000'}


In [110]:
missing_population = missing_location_rows["population"].sum()

total_population = analysis_df["population"].sum()

missing_percentage = (missing_population/ total_population * 100)

print("Population in unmatched postcodes:", missing_population)
print(
    "Percentage of Swiss population:",
    round(missing_percentage, 4),
    "%"
)

Population in unmatched postcodes: 7356
Percentage of Swiss population: 0.0806 %


only Percentage of Swiss population: 0.0806 % data is missing

In [111]:
# Create a data-quality flag for missing values
analysis_df["complete_location_data"] = (
    analysis_df[
        [
            "avg_income_per_taxpayer",
            "municipality_name",
            "canton_code",
            "postcode_latitude",
            "postcode_longitude"
        ]
    ]
    .notna()
    .all(axis=1)
)

In [112]:
analysis_df["complete_location_data"].value_counts()

complete_location_data
True     3170
False       6
Name: count, dtype: int64

In [113]:
# Create some useful indicators
analysis_df["has_migros"] = (analysis_df["migros_count"] > 0)

analysis_df["population_per_migros"] = (analysis_df["population"] / analysis_df["migros_count"].replace(0, np.nan))

analysis_df["population_per_migros"] = (analysis_df["population_per_migros"].fillna(analysis_df["population"]))


In [114]:
analysis_df.isna().sum()

postal_code                  0
population                   0
avg_income_per_taxpayer      6
municipality_name            6
canton_code                  6
aldi_count                   0
coop_count                   0
denner_count                 0
lidl_count                   0
migros_count                 0
external_competitor_count    0
migros_group_count           0
postcode_latitude            6
postcode_longitude           6
locality_name                6
complete_location_data       0
has_migros                   0
population_per_migros        0
dtype: int64

In [115]:
# save the final dataset
analysis_df.to_csv(
    "data/migros_master_data.csv",
    index=False
)

print("Final analysis shape:", analysis_df.shape)
analysis_df.head()

Final analysis shape: (3176, 18)


,postal_code,population,avg_income_per_taxpayer,municipality_name,canton_code,aldi_count,coop_count,denner_count,lidl_count,migros_count,external_competitor_count,migros_group_count,postcode_latitude,postcode_longitude,locality_name,complete_location_data,has_migros,population_per_migros
0,1000,4248,57188.176061,Lausanne,VD,0,0,0,0,0,0,0,46.552076,6.686622,Lausanne 25,True,False,4248.000000
1,1003,6879,57188.176061,Lausanne,VD,2,1,1,2,2,5,3,46.520577,6.631877,Lausanne,True,True,3439.500000
2,1004,31463,57188.176061,Lausanne,VD,1,5,1,0,3,6,4,46.528291,6.618649,Lausanne,True,True,10487.666667
3,1005,12454,57188.176061,Lausanne,VD,0,3,0,0,0,3,0,46.520897,6.642754,Lausanne,True,False,12454.000000
4,1006,15621,57188.176061,Lausanne,VD,0,2,2,0,1,2,3,46.512854,6.633803,Lausanne,True,True,15621.000000


In [116]:
# validating the dataset
print("Rows:", len(analysis_df))
print("Unique postcodes:", analysis_df["postal_code"].nunique())

print("\nMissing values:")
print(analysis_df.isna().sum())

print(
    "\nPopulation represented:",
    f"{analysis_df['population'].sum():,}"
)

print("\nStore totals:")
print(analysis_df[store_count_columns].sum())

print(
    "\nOriginal store totals:"
)
print(stores_df["store_group"].value_counts())

Rows: 3176
Unique postcodes: 3176

Missing values:
postal_code                  0
population                   0
avg_income_per_taxpayer      6
municipality_name            6
canton_code                  6
aldi_count                   0
coop_count                   0
denner_count                 0
lidl_count                   0
migros_count                 0
external_competitor_count    0
migros_group_count           0
postcode_latitude            6
postcode_longitude           6
locality_name                6
complete_location_data       0
has_migros                   0
population_per_migros        0
dtype: int64

Population represented: 9,127,125

Store totals:
migros_count    717
coop_count      923
denner_count    724
aldi_count      240
lidl_count      191
dtype: int64

Original store totals:
store_group
Coop      923
Denner    724
Migros    717
Aldi      240
Lidl      191
Name: count, dtype: int64
